plugins/backstage-plugin-devcontainers-backend/src/processors/DevcontainersProcessor.test.ts

```diff
@@ -0,0 +1,233 @@
+            url: mockUrlRoot,
+          },
+        ],
+      };
+    }),
+  } as const satisfies UrlReader;
+
+  const processor = new DevcontainersProcessor(mockReader, {
+    tagName,
+    logger: getVoidLogger(),
+  });
+
+  return { mockReader, processor } as const;
+}
+
+describe(`${DevcontainersProcessor.name}`, () => {
+  describe('getProcessorName', () => {
+    it('Should use Coder prefix in the output', () => {
+      const { processor } = setupProcessor();
+      const name = processor.getProcessorName();
+      expect(name).toMatch(new RegExp(`^${PROCESSOR_NAME_PREFIX}`));
+    });
+  });
+
+  describe('preProcessEntity', () => {
+    it('Returns unmodified entity whenever kind is not "Component"', async () => {
+      /**
+       * Formats taken from Backstage docs
+       * @see {@link https://backstage.io/docs/features/software-catalog/descriptor-format/}
+       */
+      const otherEntityKinds: readonly string[] = [
+        'Template',
+        'API',
+        'Group',
+        'User',
+        'Resource',
+        'System',
+        'Domain',
+        'Location',
+      ];
+
+      const { processor } = setupProcessor();
+      await Promise.all(
+        otherEntityKinds.map(async kind => {
+          const inputEntity = { ...baseEntity, kind };
+          const inputSnapshot = structuredClone(inputEntity);
+
+          const outputEntity = await processor.preProcessEntity(
+            inputEntity,
+            baseLocation,
+            jest.fn(),
+          );
+
+          expect(outputEntity).toBe(inputEntity);
+          expect(outputEntity).toEqual(inputSnapshot);
+        }),
+      );
+    });
+
+    it('Returns an unmodified component entity when location is not for catalog-info.yaml file', async () => {
+      const invalidLocation: LocationSpec = {
+        ...baseLocation,
+        target: 'https://www.definitely-not-valid.com/fake-repo/cool.html',
+      };
+
+      const { processor } = setupProcessor();
+      const inputSnapshot = structuredClone(baseEntity);
+
+      const outputEntity = await processor.preProcessEntity(
+        baseEntity,
+        invalidLocation,
+        jest.fn(),
+      );
+
+      expect(outputEntity).toBe(baseEntity);
+      expect(outputEntity).toEqual(inputSnapshot);
+    });
+
+    it("Produces a new component entity with the devcontainers tag when the entity's repo matches the devcontainers pattern", async () => {
+      const inputEntity = { ...baseEntity };
+      const inputSnapshot = structuredClone(inputEntity);
+      const { processor, mockReader } = setupProcessor();
+
+      const outputEntity = await processor.preProcessEntity(
+        inputEntity,
+        baseLocation,
+        jest.fn(),
+      );
+
+      expect(mockReader.readUrl).toHaveBeenCalled();
+
+      expect(outputEntity).not.toEqual(inputEntity);
+      expect(inputEntity.metadata.tags).toEqual(inputSnapshot.metadata.tags);
+      expect(inputEntity).toEqual(inputSnapshot);
```

Comment:
Yeah, these are checks to make sure nothing wasn't modified. I originally had comments here, but thought they might be overkill. I'll add them back in, just in case

Reply:
And yeah, you're right – the check is fully recursive. Will remove the extra case

Reply:
I think normally the comments would be overkill but I was going into the review with the expectation that we just need to check that the tags were added, so I was not thinking about also testing that the object was unmodified.  Another way of doing it I think would be two separate tests like:

```
it("adds tag")
it("does not modify object")
```

or something like that.